### DISCUSSION TITLE
does SAM2DynamicInteractivePredictor use memory bank for sure?

### DISCUSSION THREAD

**Discussion by baibaidj** [ultralytics/ultralytics#22151] [Q&A]
I find the SAM apis provided by this repo is really helpful in experimenting auto-labeling images and videos. 
I'm happy to see that with a box prompt of gland on a ultrasound image the SAM model can make quite good segmentation mask like this. 

<img width="977" height="697" alt="截屏2025-09-19 18 32 11" src="https://github.com/user-attachments/assets/3e9f7f08-5b54-4954-9e57-265b40a680e8" />
Not bad, right? This is generated by the SAM class as showing in https://docs.ultralytics.com/models/sam-2/#sam-2-prediction-examples . 

Then I had a try on the SAM2DynamicInteractivePredictor to see if the model can segment the same object on later frames with the box prompt. 

I followed the example from https://docs.ultralytics.com/models/sam-2/#dynamic-interactive-segment-and-track. After I initialized the predictor with the reference image and the box prompt on that frame, I tried to show the results, and the prediction score seemed much lower than that generated by the SAM class. 
<img width="985" height="740" alt="截屏2025-09-19 18 32 34" src="https://github.com/user-attachments/assets/883cab5e-b387-49a4-afd5-dffa2e173b14" />

Next, an image 8 frame later related to the reference image was given to the model for inference, and the result is as follows.
<img width="970" height="730" alt="截屏2025-09-19 18 32 47" src="https://github.com/user-attachments/assets/3ec410e8-acdd-4bf1-8711-5007a207fdea" />
The prediction score is still quite low comparing to the SAM class. 

So I was wondering where does the gap come from and if it should be concerning? 

I took a look at the code and found a place that I don't make sense of and might be related to the issue at hand
https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/sam/predict.py#L1942.  I read as long as a obj_idx of int instance is provided, memory bank won't be used and the returned feature is just from the current frame.  

What am I missing?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @baibaidj, thank you for your interest in Ultralytics 🚀! This is an automated response and an Ultralytics engineer will follow up shortly to help further. In the meantime, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered. You may also want to review the SAM-2 guides you referenced for additional context.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. For this specific case, please include:
- Exact code snippet used to initialize and run SAM and SAM2DynamicInteractivePredictor (with prompts, thresholds, and any post-processing)
- Model weights and version, `ultralytics` version, Python, PyTorch, CUDA/CUDNN, and OS
- Input frame sizes, any resizing/normalization, and whether the same prompts and frames are used across methods
- Logs/console output showing scores for both approaches and a brief note on how you measured them
- Confirmation that you can reproduce after upgrading to the latest `ultralytics` package (see Upgrade below)

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for the detailed context and screenshots 🙏. Once we have the MRE and environment details, we can help investigate the score gap and behavior you’re seeing with SAM2DynamicInteractivePredictor across frames.

**Reply by glenn-jocher** [reply]
Great question—yes, SAM2DynamicInteractivePredictor does use the memory bank. The line you highlighted in `track_step()` only slices the fused vision+memory tensor to a single item when you’re adding prompts for one object; the fusion with memory happens in `_prepare_memory_conditioned_features()`, which pulls encodings from `memory_bank` via `get_maskmem_enc()` and conditions the current frame on them. See the API details in the SAM2 reference for `get_maskmem_enc()` and `track_step()` and the usage guide for the dynamic predictor under Dynamic Interactive Segment and Track for how memory is added with `update_memory=True` [SAM2 reference](https://docs.ultralytics.com/reference/models/sam/predict/), [Dynamic Interactive usage](https://docs.ultralytics.com/models/sam-2/#dynamic-interactive-segment-and-track).

Quick way to verify memory is being used:
```python
from ultralytics.models.sam import SAM2DynamicInteractivePredictor

overrides = dict(task="segment", mode="predict", imgsz=1024, model="sam2.1_b.pt")
p = SAM2DynamicInteractivePredictor(overrides=overrides, max_obj_num=4)

# Add support frame to memory
p(source="frame_0008.jpg", bboxes=[[x1, y1, x2, y2]], obj_ids=[1], update_memory=True)
print(len(p.memory_bank))  # expect: 1

# Query later frame (uses memory)
results = p(source="frame_0016.jpg")
```

On the score gap: the dynamic predictor returns `object_score_logits` (object-presence logits scaled/clamped for gating) which are not calibrated or directly comparable to the per-mask scores you see from the `SAM` class; lower-looking values here are expected. If you need comparable confidence, base it on the mask logits themselves or your task metric. Also note the dynamic/video predictors default to non-overlapping masks in `postprocess()`, which can change mask logits slightly compared to single-image SAM; details are in the `postprocess` doc [postprocess behavior](https://docs.ultralytics.com/reference/models/sam/predict/#postprocess).

**Reply by baibaidj** [reply]
> Great question—yes, SAM2DynamicInteractivePredictor does use the memory bank. The line you highlighted in `track_step()` only slices the fused vision+memory tensor to a single item when you’re adding prompts for one object; the fusion with memory happens in `_prepare_memory_conditioned_features()`, which pulls encodings from `memory_bank` via `get_maskmem_enc()` and conditions the current frame on them. See the API details in the SAM2 reference for `get_maskmem_enc()` and `track_step()` and the usage guide for the dynamic predictor under Dynamic Interactive Segment and Track for how memory is added with `update_memory=True` [SAM2 reference](https://docs.ultralytics.com/reference/models/sam/predict/), [Dynamic Interactive usage](https://docs.ultralytics.com/models/sam-2/#dynamic-interactive-segment-and-track).
> 
> Quick way to verify memory is being used:
> 
> ```python
> from ultralytics.models.sam import SAM2DynamicInteractivePredictor
> 
> overrides = dict(task="segment", mode="predict", imgsz=1024, model="sam2.1_b.pt")
> p = SAM2DynamicInteractivePredictor(overrides=overrides, max_obj_num=4)
> 
> # Add support frame to memory
> p(source="frame_0008.jpg", bboxes=[[x1, y1, x2, y2]], obj_ids=[1], update_memory=True)
> print(len(p.memory_bank))  # expect: 1
> 
> # Query later frame (uses memory)
> results = p(source="frame_0016.jpg")
> ```
> 
> On the score gap: the dynamic predictor returns `object_score_logits` (object-presence logits scaled/clamped for gating) which are not calibrated or directly comparable to the per-mask scores you see from the `SAM` class; lower-looking values here are expected. If you need comparable confidence, base it on the mask logits themselves or your task metric. Also note the dynamic/video predictors default to non-overlapping masks in `postprocess()`, which can change mask logits slightly compared to single-image SAM; details are in the `postprocess` doc [postprocess behavior](https://docs.ultralytics.com/reference/models/sam/predict/#postprocess).

Thanks for the quick and detailed reply, which is quite useful for me to understand the code and mechanism. 
One followup question: it seems that SAM2DynamicInteractivePredictor won't update memory using its predictions and I was wondering why? I guess that the model's prediction may have errors and using it to update memory may corrupt the following prediction and the errors may also accumulate.  And yet, the predictions have been used to update memory in SAM2VideoPredictor, right? And users can add prompt to correct the prediction anyway.

**Reply by glenn-jocher** [reply]
You’re right—`SAM2DynamicInteractivePredictor` does not auto-encode its own free-run predictions; memory is only updated when you explicitly gate it via prompts (`update_memory=True`) to avoid drift/compounding errors. In contrast, `SAM2VideoPredictor` encodes predicted masks into memory during propagation; see how `SAM2Model.track_step()` “runs the memory encoder on the predicted mask … for use in future frames” in the [SAM2Model track_step reference](https://docs.ultralytics.com/reference/models/sam/modules/sam/#track_step) and how interactive writes happen only through [`update_memory()`](https://docs.ultralytics.com/reference/models/sam/predict/#update_memory). If you want opt‑in auto-updates with the dynamic predictor, you can promote predictions to memory when they pass your checks:

```python
# after adding obj_id=1 with a prompt on a support frame
pred_masks, _ = p.inference(im)  # im: np.ndarray or torch.Tensor
p.update_memory(obj_ids=[1], masks=pred_masks[[0]])  # gate by your own criteria
```

**Reply by Rainlv** [reply]
> Great question—yes, SAM2DynamicInteractivePredictor does use the memory bank. The line you highlighted in `track_step()` only slices the fused vision+memory tensor to a single item when you’re adding prompts for one object; the fusion with memory happens in `_prepare_memory_conditioned_features()`, which pulls encodings from `memory_bank` via `get_maskmem_enc()` and conditions the current frame on them. 

But why `_prepare_memory_conditioned_features` do not fuse the memory when `obj_idx` is set. 
```python
def _prepare_memory_conditioned_features(self, obj_idx: int | None) -> torch.Tensor:
  """Prepare memory-conditioned features for the current image state.
  
  If ``obj_idx`` is provided, features are prepared for a specific prompted object in the image. If ``obj_idx`` is
  None, features are prepared for all objects. If no memory is available, a no-memory embedding is added to the
  current vision features. Otherwise, memory from previous frames is used to condition the current vision features
  via a transformer attention mechanism.
  
  Args:
      obj_idx (int | None): The index of the object for which to prepare the features.
  
  Returns:
      pix_feat_with_mem (torch.Tensor): The memory-conditioned pixel features.
  """
  if len(self.memory_bank) == 0 or isinstance(obj_idx, int):
      # For initial conditioning frames, encode without using any previous memory.
      # Directly add the no-memory embedding (instead of using the transformer encoder).
      pix_feat_with_mem = self.vision_feats[-1] + self.model.no_mem_embed
  else:
      # for inference frames, use the memory features from previous frames
      memory, memory_pos_embed = self.get_maskmem_enc()
      pix_feat_with_mem = self.model.memory_attention(
          curr=self.vision_feats[-1:],
          curr_pos=self.vision_pos_embeds[-1:],
          memory=memory,
          memory_pos=memory_pos_embed,
          num_obj_ptr_tokens=0,  # num_obj_ptr_tokens
      )
  ...
```
